### PR TITLE
[3.7] #5382: Deprecate overloads of Kokkos::sort() taking parameter 'bool always_use_kokkos_sort'

### PR DESCRIPTION
**Pull request by PhilMiller**
A richer sorting interface will be implemented later, probably involving a handle class of some sort. This is just clearing out an obsolete design.

### CODE REVIEW

File: algorithms/src/Kokkos_Sort.hpp [hunk 1]
```diff
@@ -628,7 +624,31 @@ std::enable_if_t<Kokkos::is_execution_space<ExecutionSpace>::value> sort(
...
+}
+
+template <class ViewType>
+void sort(ViewType const& view) {
+  typename ViewType::execution_space exec;
+  // FIXME: Do this and the following deprecated overload need to call
+  // Kokkos::fence() first?
+  sort(exec, view);
+  exec.fence("Kokkos::Sort: fence after sorting");
+}
```
Comment (masterleinad) [inline algorithms/src/Kokkos_Sort.hpp:647]: I think it's OK to fix/change that later in `develop` (where I already had the same comment).

**Review by masterleinad — COMMENTED** [review]
Guard the deprecated functions with `KOKKOS_ENABLE_DEPRECATED_CODE_3` so we can (discuss to) remove them in `develop`.

**Comment by PhilMiller** [comment]
> Guard the deprecated functions with `KOKKOS_ENABLE_DEPRECATED_CODE_3` so we can (discuss to) remove them in `develop`.

Ah, right, thanks

File: algorithms/src/Kokkos_Sort.hpp [hunk 2]
```diff
@@ -628,7 +624,31 @@ std::enable_if_t<Kokkos::is_execution_space<ExecutionSpace>::value> sort(
...
+  } else {
+    sort(exec, view);
+  }
+}
+
+template <class ViewType>
+void sort(ViewType const& view) {
+  typename ViewType::execution_space exec;
+  // FIXME: Do this and the following deprecated overload need to call
+  // Kokkos::fence() first?
```
Comment (dalg24) [inline algorithms/src/Kokkos_Sort.hpp:644]: I don't think so

Comment (dalg24) [reply algorithms/src/Kokkos_Sort.hpp:644]: Phil pointed out to me that the data in the few might be filled by some other execution space type

Comment (PhilMiller) [reply algorithms/src/Kokkos_Sort.hpp]: https://github.com/kokkos/kokkos/issues/5385 To consider/address this

**Review by crtrott — APPROVED** [review]

**Comment by PhilMiller** [comment]
Oh, I should add a note to the changelog about this late deprecation

**Review by masterleinad — APPROVED** [review]

File: algorithms/src/Kokkos_Sort.hpp [hunk 3]
```diff
@@ -628,12 +624,38 @@ std::enable_if_t<Kokkos::is_execution_space<ExecutionSpace>::value> sort(
...
+  typename ViewType::execution_space exec;
+  sort(exec, view);
+  exec.fence("Kokkos::Sort: fence after sorting");
+}
+
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_3
+template <class ViewType>
+KOKKOS_DEPRECATED_WITH_COMMENT(
+    "Use the overload not taking bool always_use_kokkos_sort")
 void sort(ViewType const& view, bool const always_use_kokkos_sort = false) {
```
Comment (masterleinad) [inline algorithms/src/Kokkos_Sort.hpp:653]:
```suggestion
void sort(ViewType const& view, bool const always_use_kokkos_sort) {
```

**Comment by PhilMiller** [comment]
There's a bit more that needs deprecation-guarding in the test. Fixing now.

**Pull request merged**